File: mu_wang/9_17/Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        length = 0
        rst = head
        while rst:
            rst = rst.next
            length += 1     #count how long the linked list is
        if length ==1:
            return 
        a = 0   #find current location
        temp = head
        if length == n:   # if length equals n, then remove first
            temp=head.next
        else:
            while a < length-n-1:  
                a +=1
                head = head.next
            head.next = head.next.next
        return temp

File: Olivia/Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
       if(n<=0){
           return null;
       } 
       ListNode r=new ListNode(0);
       r.next=head;
       ListNode first=r;
       for(int i=0;i<n;i++){
           if(head==null){
               return null;
           }
           head=head.next;
       }
       while(head!=null){
           head=head.next;
           first=first.next;
       }
       first.next=first.next.next;
       return r.next;
       
    }
}
